File: Python3/Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height or len(height) < 3:
            return 0
        volume = 0
        left, right = 0, len(height) - 1
        l_max, r_max = height[left], height[right]
        while left < right:
            l_max, r_max = max(height[left], l_max), max(height[right], r_max)
            print(str(left) + "       " + str(right) + "         " + str(l_max) + "       " + str(r_max) + "   volume:  " + str(volume))
            if l_max <= r_max:
                volume += l_max - height[left]
                left += 1
            else:
                volume += r_max - height[right]
                right -= 1

        return volume

File: Golang/Trapping Rain Water.go
func trap(height []int) int {
	left, right := 0, len(height)-1
	leftMax, rightMax, ans := 0, 0, 0
	for left <= right {
		if height[left] < height[right] {
			if leftMax > height[left] {
				ans = ans + (leftMax - height[left])
			} else {
				leftMax = height[left]
			}
			left++
		} else {
			if rightMax > height[right] {
				ans = ans + (rightMax - height[right])
			} else {
				rightMax = height[right]
			}
			right--
		}
	}
	return ans
}